keras_core/distribute/distribution.py

```diff
@@ -0,0 +1,144 @@
+
+    This is aligned with `jax.sharding.Mesh` and `tf.dtensor.Mesh`, which
+    represents the computation devices in the global context.
+
+    See more details in
+    https://jax.readthedocs.io/en/latest/jax.sharding.html#jax.sharding.Mesh
+    and https://www.tensorflow.org/api_docs/python/tf/experimental/dtensor/Mesh.
+    """
+
+    def __init__(
+        self,
+        shape: list | tuple,
+        axis_names: list[str],
+        devices: list[str] = None,
+    ):
+        """Initialize the DeviceMesh for the given topology.
+
+        Args:
+            shape: the shape of the overall DeviceMesh, e.g. (8,) for a data
+                parallel only distribution, or (4, 2) for a model+data parallel
+                distribution.
+            axis_names: The logical name of the each axis for the DeviceMesh.
+                The length of the `axis_names` should match to the rank of the
+                `shape`. The `axis_names` will be used to match/create the
+                `TensorLayout` when distribute the data and weights.
+            devices: Optional list of devices. Default to all the available
+                devices locally from `list_devices()`.
+        """
+        pass
+
+
+class TensorLayout:
+    """The layout of a Tensor.
+
+    This is aligned with `jax.sharding.NamedSharding` and `tf.dtensor.Layout`,
+    which allocate the tensor to its logic axis based on the `DeviceMesh`. With
+    `DeviceMesh` and `TensorLayout`, the actual mapping between a Tensor to the
+    physical devices can be determined.
+
+    See more details in
+    https://jax.readthedocs.io/en/latest/jax.sharding.html#jax.sharding.NamedSharding
+    and https://www.tensorflow.org/api_docs/python/tf/experimental/dtensor/Layout.  # noqa: E501
+    """
+
+    def __init__(self, layout_spec: list | tuple):
+        """Initialize the TensorLayout with layout_spec.
+
+        Args:
+            layout_spec: list of strings that should map to the `axis_names` in
+                `DeviceMesh`. For any dimentions that doesn't need any sharding,
+                A `None` can be used a placeholder.
+        """
+        pass
+
+
+class Distribution:
+    """Base class for the distribution.
+
+    The `Distribution` has following key functionalities.
+
+    1. Distribute the model variables to the `DeviceMesh`.
+    2. Distribute the input data to the `DeviceMesh`.
+
+    It can create a context scope so that the framework to properly detect the
+    `Distribution` and distribute the variable/data accordingly.
+    """
+
+    def __init__(self, device_mesh: DeviceMesh):
+        pass
+
+    def distribute_data(self, data) -> Any:
+        """Shard the input data based on the Distribution setting.
+
+        Args:
+            data: input data, can be Tensor and np.Array.
+
+        Returns:
+            distributed data tensor.
+        """
+        pass
+
+    def distribute_variable(self, variable: KerasVariable) -> KerasVariable:
```

Comment:
We can't clone variables. Sharding is something that the variable will have to do itself in `_initialize`. So we just need to provide the info that the variable will need to query, e.g. `get_variable_layout(path)`

Reply:
I think this is not cloning the KerasVariable, but initialize/redistribute it. In the model parallel setting, I would expect the implementation to be:

```
path = variable.get_path()
layout = self.layout_map.get_variable_layout(path)
distributed_value = jax.device_put(variable.value, layout)
variable._direct_assign(distributed_value)
```

As the public API, I think its probably better to not include the layout_map in the method siguature, and leave it as an implementation detail for ModelParallelDistribution.